### ISSUE TITLE
Cannot compile antlr4-demo with gcc5.4

### ISSUE THREAD

**Issue by hrjaco** [DanMcLaughlin/antlr4#43]
I am using gcc5.4 on ubuntu 12.04.

I downloaded the ANTLR4 jar with C++ support from http://www.soft-gems.net/index.php/tools/49-the-antlr4-c-target-is-here. I was following https://github.com/DanMcLaughlin/antlr4/tree/master/runtime/Cpp#build--usage-notes to build.

I can build the antlr4_static and GenerateParser
```
Linking CXX static library ../../dist/libantlr4-runtime.a
[ 98%] Built target antlr4_static
Scanning dependencies of target GenerateParser
[ 98%] Built target GenerateParser
Scanning dependencies of target antlr4-demo
```

However, I then got complication error when starting to build antlr4-demo:
```
[ 98%] Building CXX object demo/CMakeFiles/antlr4-demo.dir/Linux/main.cpp.o
In file included from /home/jiacheng/code/antlr4/runtime/Cpp/runtime/src/atn/LexerAction.h:34:0,
                 from /home/jiacheng/code/antlr4/runtime/Cpp/runtime/src/atn/ATNDeserializer.h:34,
                 from /home/jiacheng/code/antlr4/runtime/Cpp/runtime/src/antlr4-runtime.h:84,
                 from /home/jiacheng/code/antlr4/runtime/Cpp/demo/Linux/main.cpp:10:
/home/jiacheng/code/antlr4/runtime/Cpp/runtime/src/atn/LexerActionType.h:76:3: warning: type attributes ignored after type is already defined [-Wattributes]
   };
   ^
In file included from /home/jiacheng/code/antlr4/runtime/Cpp/runtime/src/atn/ParserATNSimulator.h:34:0,
                 from /home/jiacheng/code/antlr4/runtime/Cpp/runtime/src/antlr4-runtime.h:125,
                 from /home/jiacheng/code/antlr4/runtime/Cpp/demo/Linux/main.cpp:10:
/home/jiacheng/code/antlr4/runtime/Cpp/runtime/src/atn/PredictionMode.h:106:3: warning: type attributes ignored after type is already defined [-Wattributes]
   };
   ^
In file included from /home/jiacheng/code/antlr4/runtime/Cpp/demo/Linux/main.cpp:11:0:
/home/jiacheng/code/antlr4/runtime/Cpp/demo/generated/TLexer.h:64:16: error: ‘virtual void antlrcpptest::TLexer::action(Ref<antlr4::RuleContext>&, int, int)’ marked ‘override’, but does not override
   virtual void action(Ref<RuleContext> const& context, int ruleIndex, int actionIndex) override;
                ^
/home/jiacheng/code/antlr4/runtime/Cpp/demo/generated/TLexer.h:65:16: error: ‘virtual bool antlrcpptest::TLexer::sempred(Ref<antlr4::RuleContext>&, int, int)’ marked ‘override’, but does not override
   virtual bool sempred(Ref<RuleContext> const& _localctx, int ruleIndex, int predicateIndex) override;
                ^
In file included from /home/jiacheng/code/antlr4/runtime/Cpp/demo/Linux/main.cpp:12:0:
/home/jiacheng/code/antlr4/runtime/Cpp/demo/generated/TParser.h:75:21: error: conflicting return type specified for ‘virtual ssize_t antlrcpptest::TParser::MainContext::getRuleIndex() const’
     virtual ssize_t getRuleIndex() const override;
                     ^
In file included from /home/jiacheng/code/antlr4/runtime/Cpp/runtime/src/ParserRuleContext.h:34:0,
                 from /home/jiacheng/code/antlr4/runtime/Cpp/runtime/src/InterpreterRuleContext.h:34,
                 from /home/jiacheng/code/antlr4/runtime/Cpp/runtime/src/antlr4-runtime.h:55,
                 from /home/jiacheng/code/antlr4/runtime/Cpp/demo/Linux/main.cpp:10:
/home/jiacheng/code/antlr4/runtime/Cpp/runtime/src/RuleContext.h:109:20: error:   overriding ‘virtual size_t antlr4::RuleContext::getRuleIndex() const’
     virtual size_t getRuleIndex() const;
                    ^
In file included from /home/jiacheng/code/antlr4/runtime/Cpp/demo/Linux/main.cpp:12:0:
/home/jiacheng/code/antlr4/runtime/Cpp/demo/generated/TParser.h:92:21: error: conflicting return type specified for ‘virtual ssize_t antlrcpptest::TParser::DivideContext::getRuleIndex() const’
     virtual ssize_t getRuleIndex() const override;
                     ^
In file included from /home/jiacheng/code/antlr4/runtime/Cpp/runtime/src/ParserRuleContext.h:34:0,
                 from /home/jiacheng/code/antlr4/runtime/Cpp/runtime/src/InterpreterRuleContext.h:34,
                 from /home/jiacheng/code/antlr4/runtime/Cpp/runtime/src/antlr4-runtime.h:55,
                 from /home/jiacheng/code/antlr4/runtime/Cpp/demo/Linux/main.cpp:10:
/home/jiacheng/code/antlr4/runtime/Cpp/runtime/src/RuleContext.h:109:20: error:   overriding ‘virtual size_t antlr4::RuleContext::getRuleIndex() const’
     virtual size_t getRuleIndex() const;
                    ^
...
````

Any idea on how to solve this?

**Comment by hrjaco** [comment]
Looking more into this, it seems there is an inconsistency between the function signature between the generated parser vs the runtime library, and gcc5.0 thinks they do not match:

antlr4/runtime/Cpp/demo/generated/TLexer.h
```
virtual void action(Ref<RuleContext> const& context, int ruleIndex, int actionIndex) override;
```

antlr4/runtime/Cpp/runtime/src/Recognizer.h
```
virtual void action(RuleContext *localctx, size_t ruleIndex, size_t actionIndex);
```

Any suggestion on how to solve this? I am blocked by this. A quick response would be appreciated!

**Comment by hrjaco** [comment]
Problem is solved by using my own compiled antlr4-4.5.4-SNAPSHOT.jar instead of the one linked from http://www.soft-gems.net/index.php/tools/49-the-antlr4-c-target-is-here.

Also I need to modify Linux/main.cpp to change this line to:
```
tree::ParseTree* tree = parser.main();
```

**Comment by mike-lischke** [comment]
Thanks for the reminder. Yes, it's about time to update all packages related to ANTLR4
